gnmi/bgp/bgpconfig.go

```diff
@@ -0,0 +1,489 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright (C) 2016 Nippon Telegraph and Telephone Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bgp contains a BGP implementation (using GoBGP) whose API is OpenConfig.
+package bgp
+
+import (
+	"reflect"
+
+	"golang.org/x/net/context"
+
+	"github.com/golang/glog"
+
+	"github.com/openconfig/lemming/gnmi/oc"
+	api "github.com/wenovus/gobgp/v3/api"
+	"github.com/wenovus/gobgp/v3/pkg/apiutil"
+	"github.com/wenovus/gobgp/v3/pkg/bgpconfig"
+	"github.com/wenovus/gobgp/v3/pkg/log"
+	"github.com/wenovus/gobgp/v3/pkg/packet/bgp"
+	"github.com/wenovus/gobgp/v3/pkg/server"
+	"github.com/wenovus/gobgp/v3/pkg/table"
+
+	apb "google.golang.org/protobuf/types/known/anypb"
+)
+
+// NOTE: This file is copied and modified from an internal package in GoBGP and
+// may be removed due to the following TODO. Reconsider significant
+// modifications to this code.
+//
+// TODO(wenbli): When the BGP implementation is more mature, consider skipping
```

Comment:
is there a reason not to do this now? are we in a rush to implement this?

i'm hesitant about depending on a fork in your personal github. what advantages does this implementation provide over the existing one? 
might be writing a short doc explaining

Reply:
I'm suggesting doing this right now because the OpenConfig -> GoBGP config conversion (see `intendedToGoBGP()`) is easier to do than a direct OpenConfig -> gRPC API conversion due to the existence of InitialConfig/UpdateConfig (in the copied code), and canonical GoBGP config -> gRPC API helpers in this file: https://github.com/osrg/gobgp/blob/master/internal/pkg/config/util.go#L437

This means future feature additions to BGP only has to modify `intendedToGoBGP()` and that's it, instead of having to examine how GoBGP is handling such configuration and making the necessary API struct conversions and calls to make sure we're doing the right thing for each feature. The cost is we have to keep in sync with GoBGP, but this part of the codebase isn't experiencing a high churn (https://github.com/osrg/gobgp/commits/master/internal/pkg/config, https://github.com/osrg/gobgp/commits/master/pkg/config), so I think it's worth the development cost savings.

Once our BGP implementation is more mature and less changes are expected, we can create our own conversion code, which would be a one-time effort to remove the GoBGP fork dependency.